sub/filter.c

```diff
@@ -0,0 +1,420 @@
+static void skip_speaker_label(struct sd *sd, unsigned char **rpp, unsigned char *end, unsigned char **wpp)
+{
+    int filter_harder = sd->opts->sub_filter_SDH_harder;
+    unsigned char *rp = *rpp;
+    unsigned char *wp = *wpp;
+
+    copy_ass(sd, &rp, end, &wp);
+    // copy any leading "- "
+    if (rp < end && *rp == '-') {
+        *wp++ = *rp++;
+    }
+    copy_ass(sd, &rp, end, &wp);
+    while (rp < end && *rp == ' ') {
+        *wp++ = *rp++;
+        copy_ass(sd, &rp, end, &wp);
+    }
+    // skip past valid data searching for :
+    while (rp < end && *rp != ':') {
+        copy_ass(sd, &rp, end, &wp);
+        if (rp == end ||
+            !((isalpha(*rp) && (filter_harder || isupper(*rp) || *rp == 'l')) ||
+               isdigit(*rp) ||
+               *rp == ' ' || *rp == '\'' || *rp == '#' || *rp == '.' || *rp == ',')) {
+            return;
+        }
+        rp++;
+    }
+    if (rp == end) {
+        // : was not found
+        return;
+    }
+    rp++; // skip :
+    copy_ass(sd, &rp, end, &wp);
+    if (rp == end) {
+        // end of data
+    } else if (rp < end-1 && *rp == '\\' && rp[1] == 'N') {
+        // line end follows - skip it as line is empty
+        rp += 2;
+    } else if (*rp == ' ' || filter_harder) {
+        while (rp < end && *rp == ' ') {
+            rp++;
+        }
+        if (rp < end-1 && *rp == '\\' && rp[1] == 'N') {
+            // line end follows - skip it as line is empty
+            rp += 2;
+        }
+    } else {
+        // non space follows - no speaker label
+        return;
+    }
+    *rpp = rp;
+    *wpp = wp;
+
+    return;
+}
+
+// check for bracketed text, like [SOUND]
+// and skip it while preserving ass tags
+// any characters are allowed, brackets are seldom used in normal text
+//
+// Parameters:
+//     rpp       read pointer pointer to source string, updated on return
+//     end       pointer to first byte past end of source string
+//     wpp       write pointer pointer to destination, updated on return
+//
+// scan in source string
+// the first character n source string must by the starting '['
+// and copy ass tags to destination string but
+// skipping bracketed text if it looks like SDH
+//
+// return true if bracketed text was removed.
+// if not valid SDH read and write pointers will be unchanged
+// otherwise they point to next position after text and next write position
+static bool skip_bracketed(struct sd *sd, unsigned char **rpp, unsigned char *end, unsigned char **wpp)
+{
+    unsigned char *rp = *rpp;
+    unsigned char *wp = *wpp;
+
+    rp++; // skip past '['
+    // skip past valid data searching for ]
+    while (rp < end && *rp != ']') {
+        copy_ass(sd, &rp, end, &wp);
+        if (rp == end) {
+            return false;
+        }
+        rp++;
+    }
+    if (rp == end) {
+        // ] was not found
+        return false;
+    }
+    rp++; // skip ]
+    // skip trailing spaces
+    while (rp < end && *rp == ' ') {
+        rp++;
+    }
```

Comment:
Not a null terminated string. Also should normally only be one space so the overhead would be large.